src/utils.jl

```diff
@@ -107,29 +107,88 @@ isdef(ex) = ismatch(or_(:(function _(__) _ end),
                         :(f_(__) = _)),
                     ex)
 
-function longdef(ex)
-  prewalk(ex) do ex
-    @match ex begin
-      (f_(args__) = body_) => @q function $f($(args...)) $body end
-      ((args__,) -> body_) => @q function ($(args...),) $body end
-      (arg_ -> body_) => @q function ($arg,) $body end
-      _ => ex
-    end
+function longdef1(ex)
+  @match ex begin
+    (f_(args__) = body_) => @q function $f($(args...)) $body end
+    (f_(args__)::rtype_ = body_) => @q function $f($(args...))::$rtype $body end
+    ((args__,) -> body_) => @q function ($(args...),) $body end
+    (arg_ -> body_) => @q function ($arg,) $body end
+    _ => ex
+  end
+end
+longdef(ex) = prewalk(longdef1, ex)
+
+function shortdef1(ex)
+  @match ex begin
+    function f_(args__) body_ end => @q $f($(args...)) = $body
+    function f_(args__)::rtype_ body_ end => @q $f($(args...))::$rtype = $body
+    function (args__,) body_ end => @q ($(args...),) -> $body
+    ((args__,) -> body_) => ex
+    (arg_ -> body_) => @q ($arg,) -> $body
+    _ => ex
   end
 end
+shortdef(ex) = prewalk(shortdef1, ex)
+
+""" `split_kwargs(x)` splits an argument list into positional and keyword args.
+Returns `(args::Vector, kwargs::Vector)` """
+function split_kwargs(args)
+    positional_args = []
+    kwargs = []
+    for x in args
+        if isa(x, Expr) && x.head == :parameters
+            @assert kwargs == []
+            kwargs = x.args
+        else
+            push!(positional_args, x)
+        end
+    end
+    return (positional_args, kwargs)
```

Comment:
I think this can be simplified a bit; `:parameters` will only be in the first slot, if at all. This should probably also handle expressions of the form `f(a, b = c)` correctly.